bob-backend/src/pearl/stuff.rs

```diff
@@ -129,6 +132,86 @@ impl Stuff {
         }
         start_time.timestamp().try_into().unwrap()
     }
+
+    pub(crate) async fn offload_old_filters(
+        iter: impl IntoIterator<Item = &Arc<DiskController>>,
+        limit: usize,
+    ) {
+        let now = Instant::now();
+        let mut holders = Self::collect_holders(iter).await;
+        let mut current_size = holders.iter().map(|x| x.1).sum::<usize>();
+        let initial_size = current_size;
+        if current_size < limit {
+            return;
+        }
+        holders.sort_by_key(|h| h.0.timestamp());
+        let holders_count = holders.len();
+        let mut freed = 0;
+        for (holder, size) in holders {
+            if current_size < limit {
+                break;
+            }
+            holder.offload_filter().await;
+            let new_size = holder.filter_memory_allocated().await;
+            freed += size.saturating_sub(new_size);
+            current_size = current_size.saturating_sub(size.saturating_sub(new_size));
+        }
+        let elapsed = now.elapsed();
+        if freed != 0 {
+            log::error!(
+                "Filters offloaded in {}s for {} holders: {} -> {}, {} freed",
+                elapsed.as_secs_f64(),
+                holders_count,
+                initial_size,
+                current_size,
+                freed
+            );
+        }
+    }
+
+    async fn collect_holders(
+        iter: impl IntoIterator<Item = &Arc<DiskController>>,
+    ) -> Vec<(SimpleHolder, usize)> {
+        iter.into_iter()
```

Comment:
This approach violates SOLID principle. It is better to expose `holder_iter()` function on each level, that provides iterator on holders and uses `holder_iter()` on level below. With such approach, here you can simply call top level `holder_iter()` and convert its result into `SimpleHolder`